subprojects/build-option/src/main/java/org/gradle/internal/buildoption/BuildOption.java

```diff
@@ -30,18 +31,46 @@
  */
 public interface BuildOption<T> {
 
-    @Nullable String getGradleProperty();
+    @Nullable
+    String getGradleProperty();
+
     void applyFromProperty(Map<String, String> properties, T settings);
+
     void configure(CommandLineParser parser);
+
     void applyFromCommandLine(ParsedCommandLine options, T settings);
 
     enum Origin {
-        GRADLE_PROPERTY, COMMAND_LINE;
+        GRADLE_PROPERTY {
+            @Override
+            public void handleInvalidValue(BuildOption<?> option, String value, String hint) {
+                String message = String.format("Value '%s' given for %s Gradle property is invalid%s.", value, option.getGradleProperty(), hintMessage(hint));
+                throw new IllegalArgumentException(message);
+            }
+        },
+        COMMAND_LINE {
+            @Override
+            public void handleInvalidValue(BuildOption<?> option, String value, String hint) {
+                if (!(option instanceof StringBuildOption)) {
+                    throw new IllegalStateException("Can't get command line from non StringBuildOption");
+                }
+                String commandLineOption = StringBuildOption.class.cast(option).getCommandLineOption();
+                String message = String.format("Argument value '%s' given for --%s option is invalid%s.", value, commandLineOption, hintMessage(hint));
+                throw new CommandLineArgumentException(message);
+            }
+        };
+
+        public abstract void handleInvalidValue(BuildOption<?> option, String value, String hint);
+
+        public void handleInvalidValue(BuildOption<?> option, String value) {
```

Comment:
@bmuschko I don't like too much `if/else if`s since we're using an enum already. What do you think about this usage, with `BuildOption` as the first argument? I'd like to hear your opinions.

Reply:
Generally speaking I like that the implementation is specific to the enum as you have it right now. It does avoid the if/else logic. Using the `BuildOption` parameter has drawbacks from the command line option. See comment above.